Issue #14909 - Define the 1st day of week of the calendar in config.php

saurass opened the pull request:
# Description
Fixes #14909
- Add config option to server_settings -> settings -> config.
- A form (Time_form) enables you to configure first day of the week.
- This will cause datePicker to pick selected day as first day of week.

## Approach
- In libraries/classes/Config/Forms/User/UserFormList.php added 'Config' in user form list array .
- In libraries/user_preferences.inc.php added 'Config' Tab icon.
- Made Config Form in libraries/classes/Config/Forms/User/ConfigForm.php, adding required fields.
- In libraries/classes/Config/Descriptions.php we describe input fields, form name and description.
- In libraries/config.values.php added the list of possible values for input field 'FirstDay'.
- In js/messages.php declare JS variable 'firstDayCalendar' and set its value same as that selected by user in config.
- In js/functions.js we declare 'firstDay: firstDayCalendar'.

## Result
- The datepicker shows first day of week as per selection by user in config -> time form.
- Video illustrating the fix https://drive.google.com/open?id=189VmqBHofYQNYdrY1BklT9ow0vip5pze

## WIP tasks
- Writing tests for these new features.
- Fixing lint errors.

# Contribution Checklist
Before submitting pull request, please review the following checklist:

- [x] Make sure you have read our [CONTRIBUTING.md](https://github.com/phpmyadmin/phpmyadmin/blob/master/CONTRIBUTING.md) document.
- [x] Make sure you are making a pull request against the correct branch. For example, for bug fixes in a released version use the corresponding QA branch and for new features use the _master_ branch. If you have a doubt, you can ask as a comment in the bug report or on the mailing list.
- [x] Every commit has proper `Signed-off-by` line as described in our [DCO](https://github.com/phpmyadmin/phpmyadmin/blob/master/DCO). This ensures that the work you're submitting is your own creation.
- [x] Every commit has a descriptive commit message.
- [x] Every commit is needed on its own, if you have just minor fixes to previous commits, you can squash them.
- [x] Any new functionality is covered by tests.

williamdes reviewed (approved):
LGTM
Thank you @saurass !

Maybe adding some documentation `.rst` ?
cc @ibennetch

ibennetch reviewed (changes requested):
This looks great. It's very thorough and touches on several aspects of the phpMyAdmin code.

I think you missed adding this also to `libraries/config.default.php`, where you should set the default to Sunday. Users can then use a line in `config.inc.php` or the user preferences form to change the default.

This doesn't seem to work on the calendar when in "grid edit" mode, that is where I can double-click on an entry in the Browse view in order to quickly edit, or in the Event editor. I suspect there are other instances of calling `PMA_addDatepicker` sprinkled throughout the code and we should modify all of them.

I have a few other suggestions I'll post in-line as well.

ibennetch commented on js/messages.php:
```diff
@@ -769,6 +769,9 @@ function () {
     , 'b_calendar.png' , "';\n";
 
+/* Calendar First Day */
+echo "var firstDayCalendar = '" , $GLOBALS['cfg']['FirstDay'] , "';\n";
```
...or FirstDayOfCalendar, that would be suitable as well without being too long.

saurass replied:
Yeah 'FirstDayOfCalendar' is short and elaborative. I will change it to that.

ibennetch commented:
It appears as if you have not used your real name in the signoff message. For compliance reasons, we are required to ask you for your name. You can modify your existing commit with `git commit —amend` and pushing again to your repository (it may be necessary to `—force` push), or we can rebase all of your commits in to one with a signoff, once this is complete.

If you have any questions or concerns, please let us know.

saurass commented on libraries/classes/Config/Forms/User/UserFormList.php:
```diff
@@ -27,6 +27,7 @@ class UserFormList extends BaseFormList
         'Main',
         'Import',
         'Export',
+        'Config',
```
That would be great, but in case we need to add other configurations, in various aspects, Config Form may prove useful.

Anyways I will move it to the path as you specified for now.

saurass commented on libraries/config.values.php:
```diff
@@ -189,6 +189,15 @@
+        '4' => __('Thursday'),
+        '5' => __('Friday'),
+        '6' => __('Saturday'),
+        '7' => __('Sunday'),
```
Lemme have a look at it once.

ibennetch replied:
I still think this should start on Sunday, as well.

saurass replied:
When I use 0 for 'sunday' here, Instead of saving 0 in $cfg['FirstDayOfWeek'] , it picks the week day name (Sunday, Monday etc..) instead of (0,1,2,3....) So, it dosen't work furthur. 

But yes in config.default, value by user can be set 0 to 6 (Sunday to Saturday). That is working well.

saurass commented:
Yeah I missed adding it to `libraries/config.default.php` . I will get it done

ibennetch commented:
I think the best place to add the documentation is probably the Design section: https://docs.phpmyadmin.net/en/latest/config.html#design-customization

saurass committed:
Fix #14909 - Config option for 1st day of the week for datepicker in server settings -> config -> time form

Fixes: #14909

Signed-off-by: Saurabh Srivastava <saurabhsrivastava312@gmail.com>

saurass commented:
@ibennetch @williamdes , I just made changes. Please have a look .

saurass commented:
@williamdes  @ibennetch I have added the documentations as well. Please review : )

ibennetch commented on doc/config.rst:
```diff
@@ -217,8 +217,8 @@ Basic settings
     and is a potential security hole allowing cross-frame scripting attacks or
-    clickjacking. Setting this to 'sameorigin' prevents phpMyAdmin to be 
-    included from another document in a frame, unless that document belongs 
+    clickjacking. Setting this to 'sameorigin' prevents phpMyAdmin to be
```
While we don't like extra spaces at the end of lines, this change should be part of another commit to keep the commit history clear here; otherwise one might see this in the commit log for your other work and think it's related.

It's best to revert the removal of the extra space characters.

saurass replied:
Ok, I will revert the removal of extra space

ibennetch commented on doc/config.rst:
```diff
@@ -2873,6 +2873,16 @@ Design customization
+.. config:option:: $cfg['FirstDayOfCalendar']
+
+    :type: integer
+    :default: 7
```
So it appears that the documented options for this are 0-6, but they seem to use the modulus operator to determine the number so 14 is also Sunday. That explains why both 0 and 7 are accepted values.

However, since the week starts on Sunday by most calendars, we should use the 0-6 range for Sunday-Saturday, rather than 1-7 with Monday-Sunday. It's a tradeoff between whether users will be more confused by starting counting with 0 or starting the week on Monday, and I think the right thing to do is default to 0 and change the help text to read Sunday to Saturday instead.

saurass replied:
Okay I will get that done.

ibennetch reviewed (changes requested):
I think you're really close on this, there are a few additional changes I have suggested but this is very close to being ready to merge.

saurass committed:
Default value for First Day Of Calendar to be 7 in config.default.value, i.e. Sunday

Signed-off-by: Saurabh Srivastava <saurabhsrivastava312@gmail.com>

Add documentation in config.rst for first day of calendar cfg

Signed-off-by: Saurabh Srivastava <saurabhsrivastava312@gmail.com>

Docs modification config.rst and default config value of config default to '0'

Signed-off-by: Saurabh Srivastava <saurabhsrivastava312@gmail.com>

saurass commented:
@ibennetch @williamdes , please review and suggest changes.

saurass commented:
@ibennetch @williamdes is something left ?
Please lemme know, I am eager to get it merged.

Thanks ( :

saurass commented:
@sudo-bot Deploy pr

williamdes commented:
@mauriciofauth @ibennetch I really want to merge this enhancement, it is simple and well done.
Please review and comment

@lem9 Do you have comments about this enhancement ?

mauriciofauth commented:
Merged, thanks for your contribution!

saurass commented:
Great, thanks a lot 😊

dddns commented:
Thank you @saurass and @mauriciofauth !
